### ISSUE TITLE
experimental ipvlan : container connectivity over default dummy ipvlan interface fails

### ISSUE THREAD

**Issue by mavenugo** [docker/libnetwork#1006]
```
$ sudo docker network create -d ipvlan iv
eb00df4e38c04108d61c3b1a8ed52679851da16ffcba191de965aa6d282b8063
$ sudo docker run --net=iv --name=ic4 -d busybox top
cbbacf841b17e71a32d10eb7f7c487302d6ae3573d5cfb2cec8a419e95fa957a
vagrant@vagrant-ubuntu-vivid-64 ~ $ sudo docker run --net=iv --name=ic5 -d busybox top
e64232ed35e72e330dc4bc8dd40af3383982dbd147d7e86e9079efdd3bd60451
vagrant@vagrant-ubuntu-vivid-64 ~ $ sudo docker network inspect iv
[
    {
        "Name": "iv",
        "Id": "eb00df4e38c04108d61c3b1a8ed52679851da16ffcba191de965aa6d282b8063",
        "Scope": "local",
        "Driver": "ipvlan",
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "172.19.0.0/16",
                    "Gateway": "172.19.0.1/16"
                }
            ]
        },
        "Containers": {
            "cbbacf841b17e71a32d10eb7f7c487302d6ae3573d5cfb2cec8a419e95fa957a": {
                "Name": "ic4",
                "EndpointID": "facf84644670d59f97439a0d4df88d156f07544e2f62bff9655f24a42ad58fb4",
                "MacAddress": "",
                "IPv4Address": "172.19.0.2/16",
                "IPv6Address": ""
            },
            "e64232ed35e72e330dc4bc8dd40af3383982dbd147d7e86e9079efdd3bd60451": {
                "Name": "ic5",
                "EndpointID": "8631b0f0be5424a923f58f26ca64955a0fcf0ff28a59756dafabb1837aa5125d",
                "MacAddress": "",
                "IPv4Address": "172.19.0.3/16",
                "IPv6Address": ""
            }
        },
        "Options": {}
    }
]

vagrant@vagrant-ubuntu-vivid-64 ~ $ ifconfig di-eb00df4e38c0
di-eb00df4e38c0 Link encap:Ethernet  HWaddr 2e:d5:a9:83:0e:5b
          inet6 addr: fe80::2cd5:a9ff:fe83:e5b/64 Scope:Link
          UP BROADCAST RUNNING NOARP  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:15 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:0 (0.0 B)  TX bytes:1202 (1.2 KB)

vagrant@vagrant-ubuntu-vivid-64 ~ $ sudo docker exec ic4 ifconfig
eth0      Link encap:Ethernet  HWaddr 2E:D5:A9:83:0E:5B
          inet addr:172.19.0.2  Bcast:0.0.0.0  Mask:255.255.0.0
          inet6 addr: fe80::2cd5:a9ff:fe83:e5b/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:1 errors:0 dropped:0 overruns:0 frame:0
          TX packets:9 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:92 (92.0 B)  TX bytes:734 (734.0 B)

lo        Link encap:Local Loopback
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

vagrant@vagrant-ubuntu-vivid-64 ~ $ sudo docker exec ic5 ifconfig
eth0      Link encap:Ethernet  HWaddr 2E:D5:A9:83:0E:5B
          inet addr:172.19.0.3  Bcast:0.0.0.0  Mask:255.255.0.0
          inet6 addr: fe80::2cd5:a9ff:fe83:e5b/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:1 errors:0 dropped:0 overruns:0 frame:0
          TX packets:3 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:100 (100.0 B)  TX bytes:258 (258.0 B)

lo        Link encap:Local Loopback
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

vagrant@vagrant-ubuntu-vivid-64 ~ $ sudo docker exec -it ic5 ping ic4
^Cvagrant@vagrant-ubuntu-vivid-64 ~ $ sudo docker exec -it ic5 ping 172.19.0.2
PING 172.19.0.2 (172.19.0.2): 56 data bytes
^C
--- 172.19.0.2 ping statistics ---
12 packets transmitted, 0 packets received, 100% packet loss
```

**Comment by mavenugo** [comment]
ping @nerdalert .

If the default `dummy` interface wont work for ipvlan (needs explanation), then we have to get back the original behavior to force the user to use `-o parent=ethX` when creating an ipvlan network.

**Comment by nerdalert** [comment]
Great catch, there are lots of bug reports on ipvlan in kernel v3.19. As we discussed we will plaster the caveat all over the top of the documentation rather then explicit kernel version tests or rejecting null `-o parent=` all together. 

Below are results from a kernel > 3.19 (v4.4.3) similar to the IT tests for a null parent to give some context for those who haven't tried the driver yet and interested.

- Ipvlan L2 mode network with multiple subnets without a parent specified

```
docker network  create  -d ipvlan \
	--subnet=192.168.210.0/24 \
	--subnet=192.168.212.0/24 \
	--gateway=192.168.210.1  \
	--gateway=192.168.212.1  \
	 -o ipvlan_mode=l2 ipnet212

# Start a container on each subnet
docker run --net=ipnet212 --ip=192.168.210.10 -itd alpine /bin/sh
docker run --net=ipnet212 --ip=192.168.212.10 -itd alpine /bin/sh

# Test 192.168.210.0/24 connectivity
docker run --net=ipnet212 --ip=192.168.210.9 -it --rm alpine ping -c 2 192.168.210.10

# Test 192.168.212.0/24 connectivity
docker run --net=ipnet212 --ip=192.168.212.9 -it --rm alpine ping -c 2 192.168.212.10
```
- Ipvlan L3 mode network with multiple subnets without a parent specified

```
docker network  create  -d ipvlan \
	--subnet=192.168.214.0/24 \
	--subnet=10.1.214.0/24 \
	--gateway=192.168.214.1  \
	 -o ipvlan_mode=l3 ipnet210

# Test 192.168.214.0/24 connectivity
docker run --net=ipnet210 --ip=192.168.214.10 -itd alpine /bin/sh
docker run --net=ipnet210 --ip=10.1.214.10 -itd alpine /bin/sh

# Test L3 connectivity from 10.1.214.0/24 to 192.168.212.0/24
docker run --net=ipnet210 --ip=192.168.214.9 -it --rm alpine ping -c 2 10.1.214.10

# Test L3 connectivity from 192.168.212.0/24 to 10.1.214.0/24
docker run --net=ipnet210 --ip=10.1.214.9 -it --rm alpine ping -c 2 192.168.214.10

# View the dummy interfaces created created by docker network create..
ip link | grep di-

# Stop all containers (or just the two created in the tests)
docker rm -f `docker ps -qa`

# Delete the two networks
docker network rm ipnet210 && docker network rm ipnet212

# Verify the links were removed with network deletion
ip link | grep di-
```

**FYI Note:** When the parent is empty or the `--internal` flag is used, a linux type `dummy` interface is dynamically created by Libnetwork to act as the parent. This network is completely isolated and is the equivalent to a `--internal` flag.

**Test Output**

```
(='o'=)[root@ub-151:docker$]# uname -a
Linux ub-151 4.4.3-040403-generic #201602251634 SMP Thu Feb 25 21:36:25 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

(='o'=)[root@ub-151:docker$]# docker network  create  -d ipvlan \
> --subnet=192.168.210.0/24 \
> --subnet=192.168.212.0/24 \
> --gateway=192.168.210.1  \
> --gateway=192.168.212.1  \
>  -o ipvlan_mode=l2 ipnet212
1e30973f6a20c3f02bd1fddb754e974cb3e395f1d0c46ff306fa7f2a67e7de87

(='o'=)[root@ub-151:docker$]# docker run --net=ipnet212 --ip=192.168.210.10 -itd alpine /bin/sh
1e8b11be82e939792b72be34904f1a2d25b33227051f2a5a4d785fe069f17f04

(='o'=)[root@ub-151:docker$]# docker run --net=ipnet212 --ip=192.168.212.10 -itd alpine /bin/sh
84b1d2d2c9e9c073c4d6b5b1a6c8ef2ae990c29fc025963b25c6025791850d09

(='o'=)[root@ub-151:docker$]# docker run --net=ipnet212 --ip=192.168.210.9 -it --rm alpine ping -c 2 192.168.210.10
PING 192.168.210.10 (192.168.210.10): 56 data bytes
64 bytes from 192.168.210.10: seq=0 ttl=64 time=0.115 ms
64 bytes from 192.168.210.10: seq=1 ttl=64 time=0.095 ms
--- 192.168.210.10 ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.095/0.105/0.115 ms

(='o'=)[root@ub-151:docker$]# docker run --net=ipnet212 --ip=192.168.212.9 -it --rm alpine ping -c 2 192.168.212.10
PING 192.168.212.10 (192.168.212.10): 56 data bytes
64 bytes from 192.168.212.10: seq=0 ttl=64 time=0.075 ms
64 bytes from 192.168.212.10: seq=1 ttl=64 time=0.145 ms
--- 192.168.212.10 ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.075/0.110/0.145 ms

(='o'=)[root@ub-151:docker$]# docker network  create  -d ipvlan \
> --subnet=192.168.214.0/24 \
> --subnet=10.1.214.0/24 \
> --gateway=192.168.214.1  \
>  -o ipvlan_mode=l3 ipnet210
b9b6d9a2a9205dc9a949ca102953cc6a6e06d7e126a5643233825eafdda0a657

(='o'=)[root@ub-151:docker$]# docker run --net=ipnet210 --ip=192.168.214.10 -itd alpine /bin/sh
a17bd7203ff61c58bb36fbfcd7b4c04866f87888ab5c6d12214af7132f1c0745

(='o'=)[root@ub-151:docker$]# docker run --net=ipnet210 --ip=10.1.214.10 -itd alpine /bin/sh
c27bb55cd8c17a124419a832fd235c2dca1672f020a6f431384fa71ae1139343

(='o'=)[root@ub-151:docker$]# docker run --net=ipnet210 --ip=192.168.214.9 -it --rm alpine ping -c 2 10.1.214.10
PING 10.1.214.10 (10.1.214.10): 56 data bytes
64 bytes from 10.1.214.10: seq=0 ttl=64 time=0.094 ms
64 bytes from 10.1.214.10: seq=1 ttl=64 time=0.095 ms
--- 10.1.214.10 ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.094/0.094/0.095 ms

(='o'=)[root@ub-151:docker$]# docker run --net=ipnet210 --ip=10.1.214.9 -it --rm alpine ping -c 2 192.168.214.10
PING 192.168.214.10 (192.168.214.10): 56 data bytes
64 bytes from 192.168.214.10: seq=0 ttl=64 time=0.106 ms
64 bytes from 192.168.214.10: seq=1 ttl=64 time=0.084 ms
--- 192.168.214.10 ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.084/0.095/0.106 ms

(='o'=)[root@ub-151:docker$]# ip link | grep di-
414: di-1e30973f6a20: <BROADCAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN mode DEFAULT group default
419: di-b9b6d9a2a920: <BROADCAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN mode DEFAULT group default
```

**Comment by nerdalert** [comment]
This was resolved by PR #1011

**Comment by nerdalert** [comment]
We had to go a bit higher and bump the minimum required kernel to >= 4.2 in order to resolve Ipvlan L2 mode issues. Added in #1015
